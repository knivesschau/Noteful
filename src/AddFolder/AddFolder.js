import React, {Component} from 'react';
import NotefulForm from '../NotefulForm/NotefulForm';
import notefulContext from '../notefulContext';
import config from '../config';
import ValidationError from '../ValidationError';
import './AddFolder.css';

export default class AddFolder extends Component {
   static defaultProps = {
       history: {
           push: () => { }
       },
   }
   
   updateFolderName = (name) => {
       this.setState({name}, () => {this.validateName(name)})
   }

   validateName = (folderName) => {
    const errorInfo = {...this.state.validationMessages};
    let hasError = false;

    folderName = folderName.trim();
        if (folderName.length === 0) {
        errorInfo.name = 'Please enter a name for the folder.';
        hasError = true;
        } 
        else {
            if (folderName.length < 3) {
                errorInfo.name = 'Please enter a name that is at least 3 characters long.';
                hasError = true;
            } 
            else {
                errorInfo.name = '';
                hasError = false;
            }
        }

        this.setState({
            validationMessages: errorInfo,
            validName: !hasError
            },
            this.validateForm);
    }

    validateForm = () => {
        this.setState({
            validForm: this.state.validName
        });
    }

    addNewFolder = (callback) => {
        const folder = {
            name: this.state.name
        };

        fetch(`${config.API_ENDPOINT}/folders`, {
            method: 'POST',
            headers: {
                'content-type': 'application/json'
            },
            body: JSON.stringify(folder),
        })
        .then (res => {
            if (!res.ok) {
                return res.json().then(err => {
                    throw err;
                })
            }
            return res.json();
        })
        .then (newFolder => {
            callback(newFolder);
        })
        .catch (err => alert(err));
    };
    

    render () {
        return (
            <notefulContext.Consumer>
                {(context) => (
                    <section className='AddFolder'>
                        <h2>
                            Create a New Folder:
                        </h2>

                        <NotefulForm onSubmit={(e) => {
                            e.preventDefault();
                            this.addNewFolder(context.addFolder);
                            this.props.history.push(`/folders/${folderId}`)}}>
                        
                            <div className="field">

                                <label htmlFor="folder-name-input">
                                    Name
                                </label>

                                <input type="text" id="folder-name-input" onChange={e => this.updateFolderName(e.target.value)}/>
                                <ValidationError hasError={!this.state.validName} message={this.state.validationMessages.name}/>

                            </div>

                            <div className="buttons">

                                <button type="submit" disabled={!this.state.validForm}>
                                    Add Folder 
                                </button>

                            </div>

                        </NotefulForm>
                    </section>
                )}
            </notefulContext.Consumer>
        );
    }
}