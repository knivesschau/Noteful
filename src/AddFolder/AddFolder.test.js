import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import AddFolder from './AddFolder';

describe ('AddFolder component', () => {
    it ('renders the complete AddFolder form', () => {
        const wrapper = shallow(<AddFolder/>)
        expect(toJson(wrapper)).toMatchSnapshot();
    })
})