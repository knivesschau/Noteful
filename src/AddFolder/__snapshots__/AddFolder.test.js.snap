// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFolder component renders the complete AddFolder form 1`] = `
<section
  className="AddFolder"
>
  <h2>
    Create a Folder
  </h2>
  <NotefulForm
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <label
        htmlFor="folder-name-input"
      >
        Name of Folder:
      </label>
      <input
        id="folder-name-input"
        name="folder-name"
        type="text"
      />
    </div>
    <div
      className="buttons"
    >
      <button
        type="submit"
      >
        Add Folder
      </button>
    </div>
  </NotefulForm>
</section>
`;
